flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaTransactionLog.java

```diff
@@ -0,0 +1,307 @@
+        final Properties copy = new Properties();
+        copy.putAll(checkNotNull(kafkaConfig, "kafkaConfig"));
+        copy.put("enable.auto.commit", false);
+        copy.put("key.deserializer", ByteArrayDeserializer.class.getName());
+        copy.put(ConsumerConfig.GROUP_ID_CONFIG, UUID.randomUUID().toString());
+        copy.put("auto.offset.reset", "earliest");
+        copy.put("value.deserializer", ByteArrayDeserializer.class.getName());
+        copy.put(ConsumerConfig.ISOLATION_LEVEL_CONFIG, "read_committed");
+        this.consumer = new KafkaConsumer<>(copy);
+        this.consumer.subscribe(ImmutableList.of(TRANSACTION_STATE_TOPIC_NAME));
+    }
+
+    /**
+     * This method queries Kafka's internal transaction topic and filters the transactions for the
+     * following rules.
+     * <li>transaction is in no terminal state {@link
+     *     KafkaTransactionLog#TERMINAL_TRANSACTION_STATES}
+     * <li>transactionalIdPrefix equals the one from {@link #main}
+     * <li>Either [transaction's subtaskId % {@link #numberOfParallelSubtasks} = {@link #main}'s
+     *     subtaskId] or [transaction's subtaskId == {@link #combined}'s subtaskId && transaction's
+     *     checkpointOffset >= {@link #combined}'s checkpointOffset] or [transaction's subtaskId >
+     *     {@link #minOthers} && transaction's checkpointOffset == 0]
+     *
+     * @return all transactionIds which must be aborted before starting new transactions.
+     */
+    public Set<String> getTransactionsToAbort() {
+        ConsumerRecords<byte[], byte[]> records = consumer.poll(CONSUMER_POLL_DURATION);
+        final Set<String> openTransactions = new HashSet<>();
+        while (!records.isEmpty()) {
+            records.records(TRANSACTION_STATE_TOPIC_NAME)
+                    .forEach(maybeAddTransaction(openTransactions));
+            records = consumer.poll(CONSUMER_POLL_DURATION);
+        }
+        return openTransactions;
+    }
+
+    /**
+     * Constructs a transactionalId with the following format {@code
+     * transactionalIdPrefix-subtaskId-offset}.
+     *
+     * @param transactionalIdPrefix prefix for the id
+     * @param subtaskId describing the subtask which is opening the transaction
+     * @param offset an always incrementing number usually capturing the number of checkpoints taken
+     *     by the subtask
+     * @return transactionalId
+     */
+    public static String buildTransactionalId(
+            String transactionalIdPrefix, int subtaskId, long offset) {
+        final StringBuilder sb = new StringBuilder();
+        return sb.append(transactionalIdPrefix)
+                .append(TRANSACTIONAL_ID_DELIMITER)
+                .append(subtaskId)
+                .append(TRANSACTIONAL_ID_DELIMITER)
+                .append(offset)
+                .toString();
+    }
+
+    private Consumer<ConsumerRecord<byte[], byte[]>> maybeAddTransaction(
+            Set<String> openTransactions) {
```

Comment:
```suggestion
    private void maybeAddTransaction(ConsumerRecord<byte[], byte[]> record,
            Set<String> openTransactions) {
```
and do the lambda on call-site.

Reply:
Do you mean instead of using a dedicated method to use an anonymous function?